src/frontend/flight/sync/utils.rs

```diff
@@ -0,0 +1,545 @@
+            Ok(lit(pk_col)
+                .between(lit(min_value.evaluate()?), lit(max_value.evaluate()?)))
+        })
+        .collect::<Result<Vec<Expr>>>()?
+        .into_iter()
+        .reduce(|e1: Expr, e2| e1.and(e2))
+        .unwrap())
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::frontend::flight::sync::schema::SyncSchema;
+    use crate::frontend::flight::sync::utils::compact_batches;
+    use arrow::array::{
+        BooleanArray, Float64Array, Int32Array, RecordBatch, StringArray, UInt8Array,
+    };
+    use arrow_schema::{DataType, Field, Schema};
+    use clade::sync::{ColumnDescriptor, ColumnRole};
+    use datafusion_common::assert_batches_eq;
+    use itertools::Itertools;
+    use rand::distributions::{Alphanumeric, DistString, Distribution, WeightedIndex};
+    use rand::seq::IteratorRandom;
+    use rand::Rng;
+    use std::collections::HashSet;
+    use std::sync::Arc;
+
+    #[test]
+    fn test_batch_compaction() -> Result<(), Box<dyn std::error::Error>> {
+        let schema = Arc::new(Schema::new(vec![
+            Field::new("old_c1", DataType::Int32, true),
+            Field::new("new_c1", DataType::Int32, true),
+            Field::new("value_c2", DataType::Float64, true),
+            Field::new("changed_c3", DataType::Boolean, false),
+            Field::new("value_c3", DataType::Utf8, true),
+        ]));
+
+        let column_descriptors = vec![
+            ColumnDescriptor {
+                role: ColumnRole::OldPk as i32,
+                name: "c1".to_string(),
+            },
+            ColumnDescriptor {
+                role: ColumnRole::NewPk as i32,
+                name: "c1".to_string(),
+            },
+            ColumnDescriptor {
+                role: ColumnRole::Value as i32,
+                name: "c2".to_string(),
+            },
+            ColumnDescriptor {
+                role: ColumnRole::Changed as i32,
+                name: "c3".to_string(),
+            },
+            ColumnDescriptor {
+                role: ColumnRole::Value as i32,
+                name: "c3".to_string(),
+            },
+        ];
+
+        let sync_schema = SyncSchema::try_new(column_descriptors, schema.clone())?;
+
+        // Setup the test data in row order for simpler extension
+        let (old_c1, new_c1, val_c2, chg_c3, val_c3): (
+            Vec<_>,
+            Vec<_>,
+            Vec<_>,
+            Vec<_>,
+            Vec<_>,
+        ) = vec![
+            (None, Some(1), Some(1.0), Some(false), None),
+            (None, Some(2), Some(2.0), Some(true), Some("two")),
+            (Some(1), Some(1), Some(1.1), Some(true), Some("one")),
+            (Some(2), Some(-2), Some(2.1), Some(false), None),
+            (Some(3), Some(3), Some(3.3), Some(true), Some("three")),
+            (Some(4), Some(4), Some(4.0), Some(false), Some("four")),
+            (Some(1), Some(2), Some(1.2), Some(true), None),
+            (Some(3), None, None, Some(false), None),
+        ]
```

Comment:
Ok, I extended the test a bit (it also includes a 2-hop PK change chain now).

Reply:
> (as an aside, maybe we should somehow use this partition by PK / order by changed_ts / take most recent behaviour if we know there's no PK changes in the batch since that's a simpler implementation, and then use the hashmap-based approach)

I think this would only increase complexity since we'd need to take care of both scenarios then (batches w/ and w/o PK changes). Also even then we'd need special casing for the Changed/Value not being necessarily from the last row.

Furthermore, the current mechanism also in theory enables us to support actual PK evolution (type and/or count) in future if we decide we want it.